File: coding_skill_cpp/08_strategy_pattern.cpp
/* C++ Strategy - std::function */
#include <iostream>
#include <functional>
#include <vector>
#include <algorithm>
using namespace std;

class Sorter {
    function<void(vector<int>&)> strategy;
public:
    void setStrategy(function<void(vector<int>&)> s) { strategy = s; }
    void sort(vector<int>& data) { if (strategy) strategy(data); }
};

int main() {
    cout << "\n=== C++ Strategy ===" << endl;
    Sorter sorter;
    vector<int> nums = {5, 2, 8, 1};
    
    sorter.setStrategy([](vector<int>& d) { sort(d.begin(), d.end()); });
    sorter.sort(nums);
    cout << "오름차순: "; for (int n : nums) cout << n << " "; cout << endl;
    
    sorter.setStrategy([](vector<int>& d) { sort(d.begin(), d.end(), greater<int>()); });
    sorter.sort(nums);
    cout << "내림차순: "; for (int n : nums) cout << n << " "; cout << endl;
    
    return 0;
}

File: coding_skill_py/08_strategy_pattern.py
"""Python Strategy - 일급 함수"""
from typing import Callable, List

class Sorter:
    def __init__(self):
        self.strategy: Callable = None
    
    def set_strategy(self, strategy: Callable):
        self.strategy = strategy
    
    def sort(self, data: List[int]):
        if self.strategy:
            self.strategy(data)

def main():
    print("\n=== Python Strategy ===")
    sorter = Sorter()
    nums = [5, 2, 8, 1]
    
    sorter.set_strategy(lambda d: d.sort())
    sorter.sort(nums)
    print(f"오름차순: {nums}")
    
    sorter.set_strategy(lambda d: d.sort(reverse=True))
    sorter.sort(nums)
    print(f"내림차순: {nums}")

if __name__ == "__main__":
    main()
